libraries/bot-builder/src/main/java/com/microsoft/bot/builder/skills/SkillHandlerImpl.java

```diff
@@ -0,0 +1,252 @@
+        }
+    }
+
+    private static void applySkillActivityToTurnContext(TurnContext turnContext, Activity activity) {
+        // adapter.ContinueConversation() sends an event activity with ContinueConversation in the name.
+        // this warms up the incoming middlewares but once that's done and we hit the custom callback,
+        // we need to swap the values back to the ones received from the skill so the bot gets the actual activity.
+        turnContext.getActivity().setChannelData(activity.getChannelData());
+        turnContext.getActivity().setCode(activity.getCode());
+        turnContext.getActivity().setEntities(activity.getEntities());
+        turnContext.getActivity().setLocale(activity.getLocale());
+        turnContext.getActivity().setLocalTimestamp(activity.getLocalTimestamp());
+        turnContext.getActivity().setName(activity.getName());
+        for (Map.Entry<String, JsonNode> entry : activity.getProperties().entrySet()) {
+            turnContext.getActivity().setProperties(entry.getKey(), entry.getValue());
+        }
+        turnContext.getActivity().setRelatesTo(activity.getRelatesTo());
+        turnContext.getActivity().setReplyToId(activity.getReplyToId());
+        turnContext.getActivity().setTimestamp(activity.getTimestamp());
+        turnContext.getActivity().setText(activity.getText());
+        turnContext.getActivity().setType(activity.getType());
+        turnContext.getActivity().setValue(activity.getValue());
+    }
+
+    private CompletableFuture<SkillConversationReference> getSkillConversationReference(String conversationId) {
+
+        SkillConversationReference skillConversationReference;
+        try {
+            skillConversationReference = this.conversationIdFactory.getSkillConversationReference(conversationId).join();
+        } catch (NotImplementedException ex) {
+            if (this.logger != null) {
+                this.logger.warn("Got NotImplementedException when trying to call "
+                    + "GetSkillConversationReference() on the ConversationIdFactory,"
+                    + " attempting to use deprecated GetConversationReference() method instead.");
+            }
+
+            // Attempt to get SkillConversationReference using deprecated method.
+            // this catch should be removed once we remove the deprecated method.
+            // We need to use the deprecated method for backward compatibility.
+            ConversationReference conversationReference =
+                this.conversationIdFactory.getConversationReference(conversationId).join();
+            skillConversationReference = new SkillConversationReference();
+            skillConversationReference.setConversationReference(conversationReference);
+            skillConversationReference.setOAuthScope(this.getOAuthScope.get());
+        }
+
+        if (skillConversationReference == null) {
+            if (this.logger != null) {
```

Comment:
didn't know we had to validate the logger for null before 🤔

Reply:
I think it's not necessary as this is generated in the prop of the object. But, this was a copy/paste from `SkillHandler` as the `Impl` is trying to move things between places